modules/ximgproc/src/find_ellipses.cpp

```diff
@@ -1304,8 +1304,8 @@ void EllipseDetectorImpl::preProcessing(Mat1b &image, Mat1b &dp, Mat1b &dn) {
     // 2 - the pixel does belong to an edge
     for (int i = 0; i <= imgSize.height; i++) {
         int *tmpMag = magBuffer[(i > 0) + 1] + 1;
-        const short *tmpDx = (short *) (dx[i]);
-        const short *tmpDy = (short *) (dy[i]);
+        const short *tmpDx = (short *) (dx.data + dx.step * i);
+        const short *tmpDy = (short *) (dy.data + dy.step * i);
```

Comment:
> `dx.data + dx.step * i`

Manual address arithmetic should be avoided.
Use Mat::ptr() instead (use 1 argument to get row start address).

BTW, If debug checks failed - then there is out of boundary access.

---

Also above:

> malloc

manual memory management should be avoided too. Use `std::vector` / `cv::AutoBuffer` instead.